## TG Classificação ESG Leitura e Tratamento da Origem dos dados

#### Leitura da base de dados do S&P 500 diretamento do Yahoo Finance utilizando a biblioteca yfinance, que em seguida será exportado para CSV

import pandas as pd
import yfinance as yf
import time

# 1. Captura a lista de empresas do S&P 500 da Wikipedia
url = 'https://en.wikipedia.org/wiki/List_of_S%26P_500_companies'
tabela = pd.read_html(url)
sp500_df = tabela[0]
tickers = sp500_df['Symbol'].tolist()

# Corrige tickers com ponto
tickers = [ticker.replace('.', '-') for ticker in tickers]

# 2. Coleta todas as informações disponíveis de cada empresa
def get_all_info(ticker):
    try:
        info = yf.Ticker(ticker).info
        info['Ticker'] = ticker  # adiciona a identificação
        return info
    except Exception as e:
        print(f"Erro com {ticker}: {e}")
        return None

# 3. Loop para coletar os dados
dados_completos = []
for ticker in tickers:
    print(f"Coletando: {ticker}")
    data = get_all_info(ticker)
    if data:
        dados_completos.append(data)
    time.sleep(1)  # pausa para evitar bloqueio da API

# 4. Criar DataFrame com todas as colunas possíveis
df_completo = pd.DataFrame(dados_completos)

# 5. Exportar
df_completo.to_excel('sp500_dados_completos.xlsx', index=False)
df_completo.to_csv('sp500_dados_completos.csv', index=False)

print("Coleta completa finalizada.")

In [ ]:
import yfinance as yf
import pandas as pd
import time

# Lista de tickers (sua lista completa)
tickers_personalizados = [
    "ENPH", "EMN", "DPZ", "DAY", "DVA", "DRI", "ZTS", "ZBH", "ZBRA", "YUM", "XYL", "XEL", "WYNN", "WTW", "WMB", "WY",
    "WRK", "WDC", "WST", "WELL", "WFC", "WEC", "WAT", "WM", "WBD", "WMT", "WBA", "WAB", "GWW", "WRB", "VMC", "VST", "V",
    "VICI", "VTRS", "VRTX", "VZ", "VRSK", "VRSN", "VLTO", "VTR", "VLO", "UHS", "UNH", "URI", "UPS", "UAL", "UNP", "ULTA",
    "UDR", "UBER", "USB", "TSN", "TYL", "TFC", "TRMB", "TDG", "TT", "TSCO", "TJX", "TMO", "DIS", "TRV", "SO", "SHW", "MOS",
    "KR", "KHC", "SJM", "IPG", "HSY", "HIG", "EL", "COO", "CI", "SCHW", "ALL", "TXT", "TXN", "TSLA", "TER", "TFX", "TDY",
    "TEL", "TGT", "TRGP", "TPR", "TTWO", "TMUS", "TROW", "SYY", "SNPS", "SYF", "SMCI", "SYK", "STE", "STLD", "STT", "SBUX",
    "SWK", "LUV", "SOLV", "SNA", "SWKS", "SPG", "NOW", "SRE", "STX", "SLB", "SBAC", "CRM", "SPGI", "RTX", "RCL", "ROST",
    "ROP", "ROL", "ROK", "RHI", "RVTY", "RMD", "RSG", "RF", "REGN", "REG", "O", "RJF", "RL", "DGX", "PWR", "QCOM", "QRVO",
    "PHM", "PSA", "PEG", "PTC", "PRU", "PLD", "PGR", "PG", "PFG", "PPL", "PPG", "POOL", "PNC", "PNW", "PSX", "PM", "PCG",
    "PFE", "PEP", "PNR", "PYPL", "PAYC", "PAYX", "PH", "PARA", "PANW", "PKG", "PCAR", "OTIS", "ORCL", "OKE", "ON", "OMC",
    "ODFL", "OXY", "ORLY", "NXPI", "NVR", "NVDA", "NUE", "NRG", "NCLH", "NOC", "NTRS", "NSC", "NDSN", "NI", "NKE", "NEE",
    "NWS", "NWSA", "NEM", "NFLX", "NTAP", "NDAQ", "MSCI", "MSI", "MS", "MCO", "MNST", "MPWR", "MDLZ", "TAP", "MOH", "MHK",
    "MRNA", "MAA", "MSFT", "MU", "MCHP", "MGM", "MTD", "MET", "META", "MRK", "MDT", "MCK", "MCD", "MKC", "MTCH", "MA",
    "MAS", "MLM", "MMC", "MAR", "MKTX", "MPC", "MRO", "MTB", "LYB", "LULU", "LOW", "L", "LMT", "LKQ", "LYV", "LIN", "LEN",
    "LDOS", "LVS", "LW", "LRCX", "LH", "LHX", "KLAC", "KMI", "KIM", "KMB", "KEYS", "KEY", "KDP", "KVUE", "K", "JNPR", "JPM",
    "JCI", "JNJ", "JBHT", "J", "JBL", "IRM", "IQV", "INVH", "IVZ", "ISRG", "INTU", "IP", "IFF", "IBM", "ICE", "INTC", "PODD",
    "IR", "INCY", "ILMN", "ITW", "IDXX", "IEX", "HII", "HBAN", "HUM", "HUBB", "HPQ", "HWM", "HST", "HRL", "HON", "HD",
    "HOLX", "HLT", "HPE", "HES", "HSIC", "JKHY", "DOC", "HCA", "HAS", "HAL", "GS", "GL", "GPN", "GILD", "GPC", "GM", "GIS",
    "GE", "GD", "GNRC", "GEN", "GEV", "GEHC", "IT", "GRMN", "FCX", "BEN", "FOX", "FOXA", "FTV", "FTNT", "F", "FMC", "FI",
    "FE", "FSLR", "FITB", "FIS", "FDX", "FRT", "FAST", "FICO", "FDS", "FFIV", "XOM", "EXR", "EXPD", "EXPE", "EXC", "ES",
    "EVRG", "EG", "ETSY", "ESS", "EQR", "EQIX", "EFX", "EQT", "EPAM", "EOG", "ETR", "EMR", "LLY", "ELV", "EA", "EW", "EIX",
    "ECL", "EBAY", "ETN", "DD", "DUK", "DTE", "DOW", "DOV", "D", "DLTR", "DG", "DFS", "DLR", "FANG", "DXCM", "DVN", "DAL",
    "DE", "DECK", "DHR", "DHI", "CVS", "CMI", "CSX", "CCI", "CTRA", "COST", "CSGP", "CTVA", "CPAY", "GLW", "CPRT", "CEG",
    "STZ", "ED", "COP", "CAG", "CMA", "CMCSA", "CL", "CTSH", "KO", "CMS", "CME", "CLX", "CFG", "C", "CSCO", "CTAS", "CINF",
    "CHD", "CB", "CMG", "CVX", "CHTR", "CRL", "CF", "CNP", "CNC", "COR", "CE", "CDW", "CBRE", "CBOE", "CAT", "CTLT", "CARR",
    "CCL", "KMX", "CAH", "COF", "CPB", "CPT", "CZR", "CDNS", "CHRW", "BG", "BLDR", "BF.B", "BRO", "BR", "AVGO", "BMY", "BSX",
    "BXP", "BWA", "BKNG", "BA", "BX", "BLK", "BIIB", "TECH", "BIO", "BBY", "BRK-B", "BDX", "BAX", "BBWI", "BK", "BAC",
    "BALL", "BKR", "AXON", "AVY", "AVB", "AZO", "ADP", "ADSK", "ATO", "T", "AIZ", "AJG", "ANET", "ADM", "ACGL", "APTV",
    "AMAT", "AAPL", "APA", "AON", "ANSS", "ADI", "APH", "AMGN", "AME", "AMP", "AWK", "AMT", "AIG", "AXP", "AEP", "AAL",
    "AEE", "AMCR", "AMZN", "MO", "GOOG", "GOOGL", "LNT", "ALLE", "ALGN", "ARE", "ALB", "AKAM", "ABNB", "APD", "A", "AFL",
    "AES", "AMD", "ADBE", "ACN", "ABBV", "ABT", "AOS", "MMM"
]

# Função para buscar todos os dados
def get_all_info(ticker):
    try:
        info = yf.Ticker(ticker).info
        info['Ticker'] = ticker
        return info
    except Exception as e:
        print(f"Erro com {ticker}: {e}")
        return None

# Coleta
dados_empresas = []
for ticker in tickers_personalizados:
    print(f"Coletando {ticker}")
    data = get_all_info(ticker)
    if data:
        dados_empresas.append(data)
    time.sleep(1)  # evitar sobrecarga da API

# Salvar CSV
df = pd.DataFrame(dados_empresas)
df.to_csv("dados_sp500.csv", index=False)
print("Arquivo salvo como dados_empresas_personalizadas.csv")

#### Leitura do CSV gerado a partir da leitura da base do Yahoo Finance

In [1]:
import pandas as pd

# Caminho completo para o CSV
caminho_arquivo = r'C:\Users\lucas\TG - ESG\Datasets\dados_sp500.csv'

# Ler o arquivo
df_yf = pd.read_csv(caminho_arquivo)

df_yf.head()

,address1,city,state,zip,country,phone,website,industry,industryKey,industryDisp,...,lastDividendDate,earningsGrowth,dividendDate,prevName,nameChangeDate,address2,underlyingSymbol,uuid,ipoExpectedDate,industrySymbol
0,47281 Bayside Parkway,Fremont,CA,94538,United States,707 774 7000,https://www.enphase.com,Solar,solar,Solar,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,200 South Wilcox Drive,Kingsport,TN,37662,United States,423 229 2000,https://www.eastman.com,Specialty Chemicals,specialty-chemicals,Specialty Chemicals,...,1.741910e+09,0.130,1.751933e+09,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,30 Frank Lloyd Wright Drive,Ann Arbor,MI,48105,United States,734 930 3030,https://ir.dominos.com,Restaurants,restaurants,Restaurants,...,1.741910e+09,0.301,1.751242e+09,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,3311 East Old Shakopee Road,Minneapolis,MN,55425,United States,952 853 8100,https://www.dayforce.com,Software - Application,software-application,Software - Application,...,NaN,1.250,NaN,Ceridian HCM Holding Inc.,2025-05-24,NaN,NaN,NaN,NaN,NaN
4,2000 16th Street,Denver,CO,80202,United States,720-631-2100,https://www.davita.com,Medical Care Facilities,medical-care-facilities,Medical Care Facilities,...,NaN,-0.245,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


#### Vamos utilizar a função shape para verificar o formato do dataset.

In [2]:
df_yf.shape

(503, 189)

#### Como podemos ver, nosso dataset é composto de 503 amostras com 189 features cada. Vamos verificar agora o tipo de cara feature.

In [3]:
df_types = pd.DataFrame(df_yf.dtypes, columns=['Tipo de Dado'])

In [4]:
from IPython.display import display

# Forçar exibição completa
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.expand_frame_repr', False)

# Mostrar como tabela do notebook
display(df_types)

,Tipo de Dado
address1,object
city,object
state,object
zip,object
country,object
phone,object
website,object
industry,object
industryKey,object
industryDisp,object


#### Forçando todas as colunas do DF a serem minusculas

In [5]:
df_yf.columns = df_yf.columns.str.lower()

#### Lendo base com clasificação de risco ESG

In [6]:
# Classificações de risco ESG SP 500
# Caminho do seu arquivo
caminho_arquivo_esg = r'C:\Users\lucas\TG - ESG\Datasets\SP 500 ESG Risk Ratings.csv'

# Leia o CSV em um DataFrame
df_esg = pd.read_csv(caminho_arquivo_esg)

df_esg.head()

,Symbol,Name,Address,Sector,Industry,Full Time Employees,Description,Total ESG Risk score,Environment Risk Score,Governance Risk Score,Social Risk Score,Controversy Level,Controversy Score,ESG Risk Percentile,ESG Risk Level
0,ENPH,"Enphase Energy, Inc.","47281 Bayside Parkway\nFremont, CA 94538\nUnit...",Technology,Solar,"3,157","Enphase Energy, Inc., together with its subsid...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,EMN,Eastman Chemical Company,"200 South Wilcox Drive\nKingsport, TN 37662\nU...",Basic Materials,Specialty Chemicals,"14,000",Eastman Chemical Company operates as a special...,25.3,12.8,6.6,5.8,Moderate Controversy Level,2.0,50th percentile,Medium
2,DPZ,Domino's Pizza Inc.,"30 Frank Lloyd Wright Drive\nAnn Arbor, MI 481...",Consumer Cyclical,Restaurants,"6,500","Domino's Pizza, Inc., through its subsidiaries...",29.2,10.6,6.3,12.2,Moderate Controversy Level,2.0,66th percentile,Medium
3,DAY,"Dayforce, Inc.","3311 East Old Shakopee Road\nMinneapolis, MN 5...",Technology,Software - Application,"9,084","Dayforce Inc., together with its subsidiaries,...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,DVA,Davita Inc.,"2000 16th Street\nDenver, CO 80202\nUnited States",Healthcare,Medical Care Facilities,"70,000",DaVita Inc. provides kidney dialysis services ...,22.6,0.1,8.4,14.1,Moderate Controversy Level,2.0,38th percentile,Medium


#### Vendo a quantidade de linhas e features da tabela

In [7]:
df_esg.shape

(503, 15)

#### Verificando os tipos de dados

In [8]:
df_types_esg = pd.DataFrame(df_esg.dtypes, columns=['Tipo de Dado'])

In [9]:
from IPython.display import display

# Forçar exibição completa
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.expand_frame_repr', False)

# Mostrar como tabela do notebook
display(df_types_esg)

,Tipo de Dado
Symbol,object
Name,object
Address,object
Sector,object
Industry,object
Full Time Employees,object
Description,object
Total ESG Risk score,float64
Environment Risk Score,float64
Governance Risk Score,float64


#### Mantendo no Dataframe apenas colunas que serão utilizadas posteriormente

In [10]:
import pandas as pd
import re

# Renomear as colunas de interesse
df_esg_renomeado = df_esg.rename(columns={
    "Symbol": "ticker",
    "Name": "enterprise_name",
    "Sector": "enterprise_sector",
    "Full Time Employees": "full_time_employees",
    "Total ESG Risk score": "total_esg_risk_score",
    "Environment Risk Score": "environment_risk_score",
    "Governance Risk Score": "governance_risk_score",
    "Social Risk Score": "social_risk_score",
    "Controversy Level": "controversy_level",
    "Controversy Score": "controversy_score",
    "ESG Risk Percentile": "esg_risk_percentile",
    "ESG Risk Level": "esg_risk_level"
})

# Manter apenas as colunas selecionadas
colunas_de_interesse = [
    "ticker", "esg_risk_level"
]

df_esg_final = df_esg_renomeado[colunas_de_interesse]

# Opcional: remover registros sem total_esg_risk_score
#df_renomeado = df_renomeado.dropna(subset=["total_esg_risk_score"])

# Extrair o número da coluna 'esg_risk_percentile' e converter para inteiro
#df_esg_renomeado['esg_risk_percent'] = df_esg_renomeado['esg_risk_percentile'].astype(str).str.extract(r"(\d+)").astype(float).astype('Int64')

# Remover a coluna original 'esg_risk_percentile'
#df_esg_final = df_esg_renomeado.drop(columns=['esg_risk_percentile'])

# Visualizar o resultado
print(df_esg_final.head())

  ticker esg_risk_level
0   ENPH            NaN
1    EMN         Medium
2    DPZ         Medium
3    DAY            NaN
4    DVA         Medium


#### Realizando Join das bases por meio do campo ticker

In [11]:
# Realizar o join (equivalente ao inner join do SQL/Spark)
df_joined = pd.merge(df_yf, df_esg_final, on='ticker', how='inner')

In [12]:
df_joined.head()

,address1,city,state,zip,country,phone,website,industry,industrykey,industrydisp,sector,sectorkey,sectordisp,longbusinesssummary,fulltimeemployees,companyofficers,auditrisk,boardrisk,compensationrisk,shareholderrightsrisk,overallrisk,governanceepochdate,compensationasofepochdate,executiveteam,maxage,pricehint,previousclose,open,daylow,dayhigh,regularmarketpreviousclose,regularmarketopen,regularmarketdaylow,regularmarketdayhigh,payoutratio,beta,trailingpe,forwardpe,volume,regularmarketvolume,averagevolume,averagevolume10days,averagedailyvolume10day,bid,ask,bidsize,asksize,marketcap,fiftytwoweeklow,fiftytwoweekhigh,pricetosalestrailing12months,fiftydayaverage,twohundreddayaverage,trailingannualdividendrate,trailingannualdividendyield,currency,tradeable,enterprisevalue,profitmargins,floatshares,sharesoutstanding,sharesshort,sharesshortpriormonth,sharesshortpreviousmonthdate,dateshortinterest,sharespercentsharesout,heldpercentinsiders,heldpercentinstitutions,shortratio,shortpercentoffloat,impliedsharesoutstanding,bookvalue,pricetobook,lastfiscalyearend,nextfiscalyearend,mostrecentquarter,netincometocommon,trailingeps,forwardeps,enterprisetorevenue,enterprisetoebitda,52weekchange,sandp52weekchange,quotetype,currentprice,targethighprice,targetlowprice,targetmeanprice,targetmedianprice,recommendationmean,recommendationkey,numberofanalystopinions,totalcash,totalcashpershare,ebitda,totaldebt,quickratio,currentratio,totalrevenue,debttoequity,revenuepershare,returnonassets,returnonequity,grossprofits,freecashflow,operatingcashflow,revenuegrowth,grossmargins,ebitdamargins,operatingmargins,financialcurrency,symbol,language,region,typedisp,quotesourcename,triggerable,custompricealertconfidence,marketstate,shortname,longname,corporateactions,postmarkettime,regularmarkettime,exchange,messageboardid,exchangetimezonename,exchangetimezoneshortname,gmtoffsetmilliseconds,market,esgpopulated,regularmarketchangepercent,regularmarketprice,postmarketchangepercent,postmarketprice,postmarketchange,regularmarketchange,earningstimestamp,earningstimestampstart,earningstimestampend,earningscalltimestampstart,earningscalltimestampend,isearningsdateestimate,epstrailingtwelvemonths,epsforward,epscurrentyear,priceepscurrentyear,fiftydayaveragechange,fiftydayaveragechangepercent,twohundreddayaveragechange,twohundreddayaveragechangepercent,sourceinterval,exchangedatadelayedby,averageanalystrating,cryptotradeable,hasprepostmarketdata,firsttradedatemilliseconds,regularmarketdayrange,fullexchangename,averagedailyvolume3month,fiftytwoweeklowchange,fiftytwoweeklowchangepercent,fiftytwoweekrange,fiftytwoweekhighchange,fiftytwoweekhighchangepercent,fiftytwoweekchangepercent,displayname,trailingpegratio,ticker,fax,irwebsite,dividendrate,dividendyield,exdividenddate,fiveyearavgdividendyield,earningsquarterlygrowth,lastsplitfactor,lastsplitdate,lastdividendvalue,lastdividenddate,earningsgrowth,dividenddate,prevname,namechangedate,address2,underlyingsymbol,uuid,ipoexpecteddate,industrysymbol,esg_risk_level
0,47281 Bayside Parkway,Fremont,CA,94538,United States,707 774 7000,https://www.enphase.com,Solar,solar,Solar,Technology,technology,Technology,"Enphase Energy, Inc., together with its subsid...",2781.0,"[{'maxAge': 1, 'name': 'Mr. Badrinarayanan Ko...",6.0,7.0,9.0,10.0,9.0,1.746058e+09,1.735603e+09,[],86400.0,2.0,38.005,37.77,37.600,39.97,38.005,37.77,37.600,39.97,0.0000,1.801,36.366970,10.830601,10168324.0,10168324.0,5963127.0,11089990.0,11089990.0,39.54,41.33,11.0,1.0,5.201046e+09,37.59,141.63,3.654657,52.4044,75.50230,0.00,0.000000,USD,False,4.968794e+09,0.10434,126957223.0,131207000.0,16789560.0,18320382.0,1.743379e+09,1.745971e+09,0.1280,0.03228,0.93586,3.24,0.1755,131207000.0,6.180,6.414239,1.735603e+09,1.767139e+09,1.743379e+09,148484992.0,1.09,3.66,3.491,21.987,-0.693616,0.093625,EQUITY,39.64,125.0,33.0,57.66944,50.5,2.61538,hold,36.0,1.466857e+09,11.180,2.259830e+08,1.234605e+09,1.496,1.895,1.423128e+09,152.289,10.608,0.03012,0.17132,5.243710e+08,3.584797e+08,5.1

#### Vendo a quantidade de linhas e features da nova tabela gerada após o cruzamento

In [13]:
df_joined.shape

(503, 190)

#### Assim, temos categorias numéricas e categóricas. Para as numéricas, vamos utilizar a função describe pra calcular as principais estatísticas, como mínimo, máximo, média, etc. Já para os dados categóricos, vamos calcular sua cardinalidade.

In [14]:
df_joined.describe()

,fulltimeemployees,auditrisk,boardrisk,compensationrisk,shareholderrightsrisk,overallrisk,governanceepochdate,compensationasofepochdate,maxage,pricehint,previousclose,open,daylow,dayhigh,regularmarketpreviousclose,regularmarketopen,regularmarketdaylow,regularmarketdayhigh,payoutratio,beta,trailingpe,forwardpe,volume,regularmarketvolume,averagevolume,averagevolume10days,averagedailyvolume10day,bid,ask,bidsize,asksize,marketcap,fiftytwoweeklow,fiftytwoweekhigh,pricetosalestrailing12months,fiftydayaverage,twohundreddayaverage,trailingannualdividendrate,trailingannualdividendyield,enterprisevalue,profitmargins,floatshares,sharesoutstanding,sharesshort,sharesshortpriormonth,sharesshortpreviousmonthdate,dateshortinterest,sharespercentsharesout,heldpercentinsiders,heldpercentinstitutions,shortratio,shortpercentoffloat,impliedsharesoutstanding,bookvalue,pricetobook,lastfiscalyearend,nextfiscalyearend,mostrecentquarter,netincometocommon,trailingeps,forwardeps,enterprisetorevenue,enterprisetoebitda,52weekchange,sandp52weekchange,currentprice,targethighprice,targetlowprice,targetmeanprice,targetmedianprice,recommendationmean,numberofanalystopinions,totalcash,totalcashpershare,ebitda,totaldebt,quickratio,currentratio,totalrevenue,debttoequity,revenuepershare,returnonassets,returnonequity,grossprofits,freecashflow,operatingcashflow,revenuegrowth,grossmargins,ebitdamargins,operatingmargins,postmarkettime,regularmarkettime,gmtoffsetmilliseconds,regularmarketchangepercent,regularmarketprice,postmarketchangepercent,postmarketprice,postmarketchange,regularmarketchange,earningstimestamp,earningstimestampstart,earningstimestampend,earningscalltimestampstart,earningscalltimestampend,epstrailingtwelvemonths,epsforward,epscurrentyear,priceepscurrentyear,fiftydayaveragechange,fiftydayaveragechangepercent,twohundreddayaveragechange,twohundreddayaveragechangepercent,sourceinterval,exchangedatadelayedby,firsttradedatemilliseconds,averagedailyvolume3month,fiftytwoweeklowchange,fiftytwoweeklowchangepercent,fiftytwoweekhighchange,fiftytwoweekhighchangepercent,fiftytwoweekchangepercent,trailingpegratio,dividendrate,dividendyield,exdividenddate,fiveyearavgdividendyield,earningsquarterlygrowth,lastsplitdate,lastdividendvalue,lastdividenddate,earningsgrowth,dividenddate
count,4.950000e+02,496.000000,496.000000,496.000000,496.000000,496.000000,4.960000e+02,4.980000e+02,501.0,500.0,499.000000,499.000000,499.000000,499.000000,499.000000,499.000000,499.000000,499.000000,499.000000,496.000000,473.000000,499.000000,4.990000e+02,4.990000e+02,4.990000e+02,4.990000e+02,4.990000e+02,499.000000,499.000000,499.000000,499.000000,4.990000e+02,499.000000,499.000000,499.000000,499.000000,499.000000,499.000000,499.000000,4.950000e+02,499.000000,4.990000e+02,4.990000e+02,4.990000e+02,4.990000e+02,4.990000e+02,4.990000e+02,499.000000,499.000000,499.000000,499.000000,497.000000,4.990000e+02,499.000000,499.000000,4.990000e+02,4.990000e+02,4.990000e+02,4.990000e+02,499.000000,498.000000,495.000000,468.000000,499.000000,4.990000e+02,499.000000,497.00000,497.000000,497.000000,497.000000,489.000000,497.000000,4.950000e+02,495.000000,4.710000e+02,4.950000e+02,480.000000,480.000000,4.990000e+02,442.000000,499.000000,499.000000,469.000000,4.990000e+02,4.730000e+02,4.990000e+02,498.000000,499.000000,499.000000,499.000000,4.980000e+02,5.000000e+02,500.0,499.000000,499.000000,498.000000,498.000000,498.000000,499.000000,4.980000e+02,5.000000e+02,5.000000e+02,4.900000e+02,4.900000e+02,499.000000,498.000000,496.000000,496.000000,499.000000,499.000000,499.000000,499.000000,500.0,500.0,4.990000e+02,4.990000e+02,499.000000,499.000000,499.000000,499.000000,499.000000,412.000000,406.000000,406.000000,4.260000e+02,395.000000,439.000000,3.680000e+02,427.000000,4.270000e+02,437.000000,4.190000e+02
mean,5.832195e+04,5.381048,5.366935,5.429435,5.312500,5.350806,1.746058e+09,1.730279e+09,86400.0,2.0,213.707786,211.654566,210.586960,214.297915,213.707786,211.654566,210.587191,214.297785,0.736728,1.003810,

#### Como podemos ver na primeira linha da tabela acima, nem todos as features têm a mesma quantidade de dados, isto é, temos uma certa quantidade de dados nulos. Vamos calcular em porcentagem quanto cada coluna possui.

In [15]:
df_joined.isnull().sum().div(len(df_joined)).sort_values(ascending = False)

industrysymbol                       0.998012
underlyingsymbol                     0.996024
uuid                                 0.996024
ipoexpecteddate                      0.978131
namechangedate                       0.920477
prevname                             0.920477
fax                                  0.838966
address2                             0.743539
irwebsite                            0.316103
lastsplitdate                        0.268390
lastsplitfactor                      0.268390
fiveyearavgdividendyield             0.214712
dividendrate                         0.192843
dividendyield                        0.192843
displayname                          0.190855
trailingpegratio                     0.180915
dividenddate                         0.166998
exdividenddate                       0.153082
lastdividenddate                     0.151093
lastdividendvalue                    0.151093
esg_risk_level                       0.145129
earningsgrowth                    

#### Dropando colunas com muitos dados nulos, ou com entropia 0, irrelevantes para o modelo ou que podem induzir o modelo a uma determinada classificação

In [16]:
# Calcula a porcentagem de valores nulos por coluna
porcentagem_nulos = df_joined.isnull().sum().div(len(df_joined))

# Filtra colunas com menos de 30% de nulos
colunas_validas = porcentagem_nulos[porcentagem_nulos < 0.3].index

# Mantém apenas as colunas com menos de 30% de nulos
df_filtrado = df_joined[colunas_validas]

# Dropando colunas com muitos dados nulos e que não são relevantes para o modelo
#Obs: O campo companyofficers pode ser útil mas seu tratamento requer tratamentos mais avançados que podem ser feitos no futuro
colunas_para_dropar = [
        'lastsplitfactor',
        'lastsplitdate',
        'displayname',
        'dividenddate',
        'exdividenddate',
        'lastdividenddate',
        'earningscalltimestampend',
        'earningscalltimestampstart',
        'governanceepochdate',
        'zip',
        'phone',
        'compensationasofepochdate',
        'earningstimestamp',
        'corporateactions',
        'address1',
        'city',
        'website',
        'industry',
        'industrydisp',
        'sector',
        'sectordisp',
        'longbusinesssummary',
        'currency',
        'quotetype',
        'financialcurrency',
        'financialcurrency',
        'symbol',
        'language',
        'region',
        'typedisp',
        'quotesourcename',
        'messageboardid',
        'exchangetimezonename',
        'exchangetimezoneshortname',
        'market',
        'marketstate',
        'shortname',
        'longname',
        'averageanalystrating',
        'postmarkettime',
        'companyofficers',
        'executiveteam',
        'custompricealertconfidence',
        'tradeable',
        'triggerable',
        'esgpopulated',
        'cryptotradeable',
        'hasprepostmarketdata',
        'fullexchangename',
        'isearningsdateestimate',
        'state', 
        'country', 
        'industrykey',
        'sectorkey', 
        'recommendationkey', 
        'exchange',
        'pricehint',
        'maxage',
        'gmtoffsetmilliseconds',
        'sandp52weekchange',
        'sourceinterval',
        'exchangedatadelayedby',
        'dateshortinterest',
        'regularmarkettime',
        'sharesshortpreviousmonthdate',
        'open',
        'regularmarketpreviousclose',
        'firsttradedatemilliseconds',
        'earningstimestampstart',
    	'earningstimestampend',
    	'targethighprice',
    	'mostrecentquarter',
    	'targetlowprice',
    	'fiftytwoweekrange_min',
    	'regularmarketdayrange_min',
    	'ask',
    	'currentprice',
    	'daylow',
    	'fiftydayaverage',
    	'postmarketprice',
    	'dayhigh',
    	'targetmeanprice',
    	'previousclose',
    	'regularmarketprice',
    	'regularmarketopen',
        'regularmarketdaylow',
    	'regularmarketdayhigh']
    
df = df_filtrado.drop(columns=colunas_para_dropar, errors='ignore')

In [17]:
df.shape

(503, 97)

#### Deletando os 4 tickers pois estão quase sem quaisquer informações

In [18]:
# Lista dos tickers a serem removidos
tickers_para_remover = ['WRK', 'MRO', 'CTLT', 'BF.B']

# Remove as linhas cujo Ticker está na lista
df = df[~df['ticker'].isin(tickers_para_remover)]

In [19]:
df.shape

(499, 97)

#### Preenchendo valores nulos com 0 nas colunas indicadas

In [20]:
# Lista das colunas a preencher
colunas_para_preencher = [
    'fiveyearavgdividendyield',
    'dividendrate',
    'dividendyield',
    'lastdividendvalue'
]

# Preenche os valores nulos com 0 nessas colunas
df.loc[:, colunas_para_preencher] = df[colunas_para_preencher].fillna(0)

#### Preenchendo valores nulos com média e moda

In [21]:
import pandas as pd

# Dicionário com as colunas e o método de preenchimento
colunas_preenchimento = {
    'trailingpegratio': 'media',
    'earningsgrowth': 'media',
    'earningsquarterlygrowth': 'media',
    'debttoequity': 'media',
    'enterprisetoebitda': 'media',
    'returnonequity': 'media',
    'ebitda': 'media',
    'freecashflow': 'media',
    'trailingpe': 'media',
    'quickratio': 'media',
    'currentratio': 'media',
    'recommendationmean': 'media',
    'enterprisetorevenue': 'media',
    'enterprisevalue': 'media',
    'totaldebt': 'media',
    'totalcashpershare': 'media',
    'totalcash': 'media',
    'overallrisk': 'media',
    'priceepscurrentyear': 'media',
    'epscurrentyear': 'media',
    'beta': 'media',
    'auditrisk': 'moda',
    'compensationrisk': 'moda',
    'shareholderrightsrisk': 'moda',
    'boardrisk': 'moda',
    'shortpercentoffloat': 'media',
    'numberofanalystopinions': 'moda',
    'revenuegrowth': 'media',
    'operatingcashflow': 'media',
    'epsforward': 'media',
    'forwardeps': 'media',
    'targetmedianprice': 'media',
    'fulltimeemployees': 'media',
    'postmarketchangepercent': 'media',
    'postmarketchange': 'media'
    
}

# Aplica o preenchimento apropriado
for coluna, metodo in colunas_preenchimento.items():
    if metodo == 'media':
        valor = pd.to_numeric(df[coluna], errors='coerce').mean(skipna=True)
    elif metodo == 'moda':
        valor = df[coluna].mode(dropna=True)
        valor = valor.iloc[0] if not valor.empty else None
    else:
        continue  # ignora métodos não reconhecidos

    # Preenche os valores nulos com o valor calculado
    df[coluna] = df[coluna].fillna(valor)

#### Agora não temos mais valores nulos em nossas features apenas no nosso target

In [22]:
df.isnull().sum().div(len(df_joined)).sort_values(ascending = False)

esg_risk_level                       0.139165
auditrisk                            0.000000
boardrisk                            0.000000
compensationrisk                     0.000000
shareholderrightsrisk                0.000000
overallrisk                          0.000000
payoutratio                          0.000000
beta                                 0.000000
trailingpe                           0.000000
forwardpe                            0.000000
volume                               0.000000
regularmarketvolume                  0.000000
averagevolume                        0.000000
averagevolume10days                  0.000000
averagedailyvolume10day              0.000000
bid                                  0.000000
bidsize                              0.000000
asksize                              0.000000
marketcap                            0.000000
fiftytwoweeklow                      0.000000
fiftytwoweekhigh                     0.000000
pricetosalestrailing12months      

#### Tratando colunas com range de valores ('37.59 - 141.63')

In [23]:
# Trata regularmarketdayrange
df[['regularmarketdayrange_min', 'regularmarketdayrange_max']] = (
    df['regularmarketdayrange']
    .str.split(' - ', expand=True)
    .astype(float)
)

# Trata fiftytwoweekrange
df[['fiftytwoweekrange_min', 'fiftytwoweekrange_max']] = (
    df['fiftytwoweekrange']
    .str.split(' - ', expand=True)
    .astype(float)
)

# Dropando colunas originais
df = df.drop(columns=['regularmarketdayrange', 'fiftytwoweekrange'])

#### Analisando nosso target

##### Vendo a quantida de valores distintos existentes

In [24]:
df['esg_risk_level'].nunique()

5

##### Vendo quais são esses valores

In [25]:
df['esg_risk_level'].unique()

array([nan, 'Medium', 'Low', 'High', 'Severe', 'Negligible'], dtype=object)

##### Vendo a quantidade de cada valor distinto

In [26]:
df['esg_risk_level'].value_counts()

esg_risk_level
Low           187
Medium        184
High           49
Negligible      6
Severe          3
Name: count, dtype: int64

##### Transformando os clases de com pouca quantidade para melhor performance do modelo

In [27]:
df['esg_risk_level'] = df['esg_risk_level'].replace({
    'Negligible': 'Low',
    'Severe': 'High'
})

In [28]:
df['esg_risk_level'].value_counts()

esg_risk_level
Low       193
Medium    184
High       52
Name: count, dtype: int64

In [29]:
df['esg_risk_level'].unique()

array([nan, 'Medium', 'Low', 'High'], dtype=object)

#### Verificando a quantidade de nulos em nosso target

In [30]:
qtd_nulos = df['esg_risk_level'].isnull().sum()
print(f"Número de registros com esg_risk_level nulo: {qtd_nulos}")

Número de registros com esg_risk_level nulo: 70


### Salvando DF filtrado para possibilitar análise por setor

In [33]:
# Salvar CSV
df_salvar_f = pd.DataFrame(df_filtrado)
df_salvar_f.to_csv("base_filtrada.csv", index=False)
print("Arquivo salvo como base_filtrada.csv")

Arquivo salvo como base_filtrada.csv


### Lendo DF filtrado

In [1]:
import pandas as pd

# Caminho completo para o CSV
caminho_arquivo_filtrado = r'C:\Users\lucas\TG - ESG\Datasets\base_filtrada.csv'

# Ler o arquivo
df_filtrado = pd.read_csv(caminho_arquivo_filtrado)

df_filtrado.head()

,address1,city,state,zip,country,phone,website,industry,industrykey,industrydisp,...,exdividenddate,fiveyearavgdividendyield,earningsquarterlygrowth,lastsplitfactor,lastsplitdate,lastdividendvalue,lastdividenddate,earningsgrowth,dividenddate,esg_risk_level
0,47281 Bayside Parkway,Fremont,CA,94538,United States,707 774 7000,https://www.enphase.com,Solar,solar,Solar,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,200 South Wilcox Drive,Kingsport,TN,37662,United States,423 229 2000,https://www.eastman.com,Specialty Chemicals,specialty-chemicals,Specialty Chemicals,...,1.749773e+09,3.22,0.103,2:1,1.317686e+09,0.83,1.741910e+09,0.130,1.751933e+09,Medium
2,30 Frank Lloyd Wright Drive,Ann Arbor,MI,48105,United States,734 930 3030,https://ir.dominos.com,Restaurants,restaurants,Restaurants,...,1.749773e+09,1.06,0.215,NaN,NaN,1.74,1.741910e+09,0.301,1.751242e+09,Medium
3,3311 East Old Shakopee Road,Minneapolis,MN,55425,United States,952 853 8100,https://www.dayforce.com,Software - Application,software-application,Software - Application,...,NaN,NaN,1.099,NaN,NaN,NaN,NaN,1.250,NaN,NaN
4,2000 16th Street,Denver,CO,80202,United States,720-631-2100,https://www.davita.com,Medical Care Facilities,medical-care-facilities,Medical Care Facilities,...,NaN,NaN,-0.320,2:1,1.378685e+09,NaN,NaN,-0.245,NaN,Medium


### Salvando DF tratado

In [31]:
# Salvar CSV
df_salvar = pd.DataFrame(df)
df_salvar.to_csv("base_tratada.csv", index=False)
print("Arquivo salvo como base_tratada.csv")

Arquivo salvo como base_tratada.csv


### Lendo DF tratado

In [32]:
import pandas as pd

# Caminho completo para o CSV
caminho_arquivo = r'C:\Users\lucas\TG - ESG\Datasets\base_tratada.csv'

# Ler o arquivo
df = pd.read_csv(caminho_arquivo)

df.head()

,fulltimeemployees,auditrisk,boardrisk,compensationrisk,shareholderrightsrisk,overallrisk,payoutratio,beta,trailingpe,forwardpe,volume,regularmarketvolume,averagevolume,averagevolume10days,averagedailyvolume10day,bid,bidsize,asksize,marketcap,fiftytwoweeklow,fiftytwoweekhigh,pricetosalestrailing12months,twohundreddayaverage,trailingannualdividendrate,trailingannualdividendyield,enterprisevalue,profitmargins,floatshares,sharesoutstanding,sharesshort,sharesshortpriormonth,sharespercentsharesout,heldpercentinsiders,heldpercentinstitutions,shortratio,shortpercentoffloat,impliedsharesoutstanding,bookvalue,pricetobook,lastfiscalyearend,nextfiscalyearend,netincometocommon,trailingeps,forwardeps,enterprisetorevenue,enterprisetoebitda,52weekchange,targetmedianprice,recommendationmean,numberofanalystopinions,totalcash,totalcashpershare,ebitda,totaldebt,quickratio,currentratio,totalrevenue,debttoequity,revenuepershare,returnonassets,returnonequity,grossprofits,freecashflow,operatingcashflow,revenuegrowth,grossmargins,ebitdamargins,operatingmargins,regularmarketchangepercent,postmarketchangepercent,postmarketchange,regularmarketchange,epstrailingtwelvemonths,epsforward,epscurrentyear,priceepscurrentyear,fiftydayaveragechange,fiftydayaveragechangepercent,twohundreddayaveragechange,twohundreddayaveragechangepercent,averagedailyvolume3month,fiftytwoweeklowchange,fiftytwoweeklowchangepercent,fiftytwoweekhighchange,fiftytwoweekhighchangepercent,fiftytwoweekchangepercent,trailingpegratio,ticker,dividendrate,dividendyield,fiveyearavgdividendyield,earningsquarterlygrowth,lastdividendvalue,earningsgrowth,esg_risk_level,regularmarketdayrange_min,regularmarketdayrange_max,fiftytwoweekrange_min,fiftytwoweekrange_max
0,2781.0,6.0,7.0,9.0,10.0,9.0,0.0000,1.801,36.366970,10.830601,10168324.0,10168324.0,5963127.0,11089990.0,11089990.0,39.54,11.0,1.0,5.201046e+09,37.59,141.63,3.654657,75.50230,0.00,0.000000,4.968794e+09,0.10434,126957223.0,131207000.0,16789560.0,18320382.0,0.1280,0.03228,0.93586,3.24,0.1755,131207000.0,6.180,6.414239,1.735603e+09,1.767139e+09,148484992.0,1.09,3.66,3.491,21.987,-0.693616,50.5,2.61538,36.0,1.466857e+09,11.180,2.259830e+08,1.234605e+09,1.496,1.895,1.423128e+09,152.289000,10.608,0.03012,0.17132,5.243710e+08,3.584797e+08,5.129060e+08,0.352,0.36846,0.15879,0.09853,4.302061,0.277499,0.110001,1.634998,1.09,3.66,2.44745,16.196450,-12.764401,-0.243575,-35.862297,-0.474983,5963127.0,2.049999,0.054536,-101.990005,-0.720116,-69.361565,0.477600,ENPH,0.00,0.00,0.00,0.249626,0.00,0.226513,NaN,37.600,39.97,37.59,141.63
1,14000.0,7.0,8.0,7.0,6.0,7.0,0.4178,1.304,10.259872,9.257472,1056459.0,1056459.0,1429998.0,1417400.0,1417400.0,79.98,8.0,9.0,9.299149e+09,70.90,114.50,0.993287,95.42480,3.28,0.040315,1.406207e+10,0.09848,114438757.0,115460000.0,3198611.0,2944720.0,0.0277,0.00756,0.93054,1.75,0.0321,115460000.0,50.871,1.583220,1.735603e+09,1.767139e+09,922000000.0,7.85,8.70,1.502,7.313,-0.200437,92.5,1.85000,20.0,4.180000e+08,3.622,1.923000e+09,5.112000e+09,0.739,1.719,9.362000e+09,86.003000,80.603,0.05952,0.16032,2.332000e+09,5.323750e+08,1.136000e+09,-0.009,0.24909,0.20540,0.13930,-1.007870,0.633228,0.510002,-0.820000,7.85,8.70,7.41324,10.864346,-0.985199,-0.012085,-14.884796,-0.155985,1429998.0,9.639999,0.135966,-33.960000,-0.296594,-20.043682,2.015500,EMN,3.32,4.12,3.22,0.103000,0.83,0.130000,Medium,79.535,80.77,70.90,114.50
2,6400.0,5.0,5.0,8.0,8.0,7.0,0.3595,1.161,27.547590,27.235262,352287.0,352287.0,653724.0,451350.0,451350.0,477.90,1.0,1.0,1.645064e+10,396.06,538.44,3.475127,446.03030,6.27,0.013001,2.248561e+10,0.12844,33254603.0,34241500.0,1592194.0,1526648.0,0.0465,0.08071,0.93183,2.22,0.0527,36813500.0,-112.880,-4.256113,1.735430e+09,1.766966e+09,607996992.0,17.44,17.64,4.750,24.048,-0.056278,530.0,1.90323,28.0,2.953520e+08,8.117,9.350310e+08,5.299668e+09,0.358,0.599,4.733825e+09,143.139624,136.902,0.30322,0.30343,1.346251e+09,4.742266e+08,6.805090e+08,0.031,0.28439,0.19752,0.18064,-0.379467,0.054120,0.260010,-1.830017,17.44,17.